File: kotlin/scrabble-score/src/main/kotlin/Scrabble.kt
object Scrabble {
    fun scoreWord(input: String): Int {
        if (input.isNullOrBlank()) {
            return 0
        }

        return input
                .map { Letter(it).score() }
                .sum()
    }

}

class Letter(private val value: Char) {
    fun score(): Int {
        return when (value) {
            'a', 'e', 'i', 'r', 's', 't', 'u' -> 1
            'f', 'y' -> 4
            'k' -> 5
            'q' -> 10
            else -> {
                throw IllegalArgumentException("invalid letter '$value'")
            }
        }
    }

}

File: haskell/scrabble-score/src/Scrabble.hs
module Scrabble
  ( scoreLetter
  , scoreWord
  ) where

import Data.Char (toLower)

scoreLetter :: Char -> Integer
scoreLetter l
  | lower `elem` "at" = 1
  | lower `elem` "f" = 4
  | lower `elem` "z" = 10
  | otherwise = 0
  where
    lower = toLower l

scoreWord :: String -> Integer
scoreWord = sum . map scoreLetter